src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java

```diff
@@ -114,57 +122,144 @@ private GetPathsResult getPath(AvailableNetwork network, Flow flow,
+        HaPathBuilder haPath = HaPath.builder();
+        if (forward) {
+            if (!firstPath.getSrcSwitchId().equals(secondPath.getSrcSwitchId())) {
+                throw new IllegalArgumentException(format("Forward HA sub paths must have equal source switch."
+                        + "The first sub path: %s, the  second sub path: %s", firstPath, secondPath));
+            }
+            haPath.sharedSwitchId(firstPath.getSrcSwitchId());
+            haPath.yPointSwitchId(findYPointForForwardPaths(firstPath, secondPath));
+        } else {
+            if (!firstPath.getDestSwitchId().equals(secondPath.getDestSwitchId())) {
+                throw new IllegalArgumentException(format("Reverse HA sub paths must have equal destination switch."
+                        + "The first sub path: %s, the  second sub path: %s", firstPath, secondPath));
+            }
+            haPath.sharedSwitchId(firstPath.getDestSwitchId());
+            haPath.yPointSwitchId(findYPointForReversePaths(firstPath, secondPath));
+        }
+
+        haPath.subPaths(Lists.newArrayList(firstPath, secondPath));
+        haPath.latency(Math.max(firstPath.getLatency(), secondPath.getLatency()));
+        haPath.isBackupPath(firstPath.isBackupPath() || secondPath.isBackupPath());
+        haPath.minAvailableBandwidth(Math.min(
+                firstPath.getMinAvailableBandwidth(), secondPath.getMinAvailableBandwidth()));
+        return haPath.build();
+    }
+
+    private static SwitchId findYPointForForwardPaths(Path firstPath, Path secondPath) {
+        int id = findLastForwardSharedSegmentId(firstPath.getSegments(), secondPath.getSegments());
+        if (id == -1) {
+            return firstPath.getSrcSwitchId();
+        } else {
+            return firstPath.getSegments().get(id).getDestSwitchId();
+        }
+    }
+
+    private static SwitchId findYPointForReversePaths(Path firstPath, Path secondPath) {
+        int id = findFirstReverseSharedSegmentId(firstPath.getSegments(), secondPath.getSegments());
+        if (id == firstPath.getSegments().size()) {
+            return firstPath.getDestSwitchId();
+        } else {
+            return firstPath.getSegments().get(id).getSrcSwitchId();
+        }
+    }
+
+    private static int findLastForwardSharedSegmentId(List<Segment> firstSegments, List<Segment> secondSegments) {
+        int i = 0;
+        while (i < firstSegments.size() && i < secondSegments.size()
+                && firstSegments.get(i).areEndpointsEqual(secondSegments.get(i))) {
+            i++;
+        }
+        return i - 1;
+    }
+
+    private static int findFirstReverseSharedSegmentId(List<Segment> firstSegments, List<Segment> secondSegments) {
+        int i = firstSegments.size() - 1;
+        int j = secondSegments.size() - 1;
+
+        while (i >= 0 && j >= 0 && firstSegments.get(i).areEndpointsEqual(secondSegments.get(j))) {
+            i--;
+            j--;
+        }
+        return i + 1;
```

Comment:
If none segments are equal these methods still return a value and it's gonna be out of range. It's not what the method name suggests and you are encoding some special meaning in -1 and size(). I'd suggest to use some different approach, like Optional

Reply:
fixed